File: solutions/haskell/rational-numbers/1/src/RationalNumbers.hs
module RationalNumbers
  ( Rational,
    abs,
    numerator,
    denominator,
    add,
    sub,
    mul,
    div,
    pow,
    expRational,
    expReal,
    rational,
  )
where

import GHC.TypeLits (Nat)
import Prelude hiding (Rational, abs, div)

-- Data definition -------------------------------------------------------------
data Rational a = Rational a a deriving (Eq, Show)

rational :: (Integral a) => (a, a) -> Rational a
rational (m, n) =
  if n < 0
    then rational (-m, -n)
    else Rational (quot m (gcd m n)) (quot n (gcd m n))

-- unary operators -------------------------------------------------------------
abs :: (Integral a) => Rational a -> Rational a
abs (Rational m n) =
  if m < 0
    then Rational (-m) n
    else Rational m n

numerator :: (Integral a) => Rational a -> a
numerator (Rational m n) = m

denominator :: (Integral a) => Rational a -> a
denominator (Rational m n) = n

-- binary operators ------------------------------------------------------------
add :: (Integral a) => Rational a -> Rational a -> Rational a
add (Rational m n) (Rational o p) = rational ((m * p + n * o), (n * p))

sub :: (Integral a) => Rational a -> Rational a -> Rational a
sub (Rational m n) (Rational o p) = rational ((m * p - n * o), (n * p))

mul :: (Integral a) => Rational a -> Rational a -> Rational a
mul (Rational m n) (Rational o p) = rational (m * o, n * p)

div :: (Integral a) => Rational a -> Rational a -> Rational a
div (Rational m n) (Rational o p) = rational (m * p, n * o)

pow :: (Integral a) => Rational a -> a -> Rational a
pow (Rational m n) p =
  if p >= 0
    then rational (m ^ p, n ^ p)
    else rational (n ^ (-p), m ^ (-p))

expRational :: (Integral a) => (Floating b) => Rational a -> b -> b
expRational (Rational m n) x = (fromIntegral m) ** x / (fromIntegral n) ** x

expReal :: (Floating a) => (Integral b) => a -> Rational b -> a
expReal x (Rational m n) = x ** ((fromIntegral m) / (fromIntegral n))

File: solutions/typescript/rational-numbers/1/rational-numbers.ts
export class Rational {
  private n: number;
  private d: number
  constructor(n: number, d: number) {
    const g = gcd(n, d);
    this.n = d*g >= 0 ? n/g : -n/g;
    this.d = d*g >= 0 ? d/g : -d/g;
  }

  get numerator(): number {
    return this.n
  }

  get denominator(): number {
    return this.d
  }

  add(other: Rational): Rational {
    return new Rational(this.n*other.d+this.d*other.n, this.d*other.d)
  }

  sub(other: Rational): Rational {
    return new Rational(this.n*other.d-this.d*other.n, this.d*other.d)
  }

  mul(other: Rational): Rational {
    return new Rational(this.n*other.n, this.d*other.d)
  }

  div(other: Rational): Rational {
    if (other.n === 0) {
      return new Rational(0, 1)
    }
    return new Rational(this.n*other.d, this.d*other.n)
  }

  abs(other: Rational): Rational {
    return new Rational(Math.abs(this.n), Math.abs(this.d))
  }

  exprational(x: number) {
    return x >= 0 ? new Rational(this.n**x, this.d**x) : new Rational(this.d**(-x), this.n**(-x))
  }

  expreal(x: number) {
    return x**(this.n/this.d)
  }

  reduce(): Rational {
    return this
  }
}

function gcd(a: number, b: number): number {
  if (Math.abs(a) < Math.abs(b)) {
    return gcd(b, a)
  }
  if (b === 0) {
    return a
  }
  return gcd(b, a%b)
}
